File: src/calculator/stripe_calculator.rs
use crate::config;

use super::{ChargeAmount, SaleBreakdown};

const PAYMENT_PROCESSING_PERCENTAGE: f64 = 0.015;
const EU_PAYMENT_PROCESSING_PERCENTAGE: f64 = 0.025;
const INTERNATIONAL_PAYMENT_PROCESSING_PERCENTAGE: f64 = 0.035;
const PAYMENT_PROCESSING_FEE: f64 = 0.2;

pub(crate) fn based_on_sale(
    sale: f64,
    delivery_costs: f64,
    eu: bool,
    international: bool,
) -> SaleBreakdown {
    let config = config::get_config();

    let sale_total = sale + delivery_costs;
    let payment_processing_percentage = if eu {
        EU_PAYMENT_PROCESSING_PERCENTAGE
    } else if international {
        INTERNATIONAL_PAYMENT_PROCESSING_PERCENTAGE
    } else {
        PAYMENT_PROCESSING_PERCENTAGE
    };
    let payment_processing_cost =
        (sale_total * payment_processing_percentage) + PAYMENT_PROCESSING_FEE;
    let tax = sale * (config.tax_rate / 100.0);
    let revenue = sale - payment_processing_cost - tax;
    let percentage_kept = (revenue / sale) * 100.0;
    let max_working_hours = revenue / config.hourly_rate;
    SaleBreakdown {
        sale,
        delivery_costs,
        transaction_cost: 0.0,
        payment_processing_cost,
        offsite_ads_cost: 0.0,
        regulatory_operating_fee: 0.0,
        vat_paid_by_buyer: 0.0,
        vat_on_seller_fees: 0.0,
        total_fees: 0.0,
        total_fees_with_vat: 0.0,
        tax,
        revenue,
        percentage_kept,
        max_working_hours,
    }
}

pub(crate) fn how_much_to_charge(
    number_of_minutes: f64,
    material_costs: f64,
    delivery_costs: f64,
    eu: bool,
    international: bool,
) -> ChargeAmount {
    let config = config::get_config();

    let base_charge =
        ((number_of_minutes / 60.0) * config.hourly_rate) + material_costs + delivery_costs;
    let payment_processing_percentage = if eu {
        EU_PAYMENT_PROCESSING_PERCENTAGE
    } else if international {
        INTERNATIONAL_PAYMENT_PROCESSING_PERCENTAGE
    } else {
        PAYMENT_PROCESSING_PERCENTAGE
    };
    let payment_processing_cost =
        (base_charge * payment_processing_percentage) + PAYMENT_PROCESSING_FEE;
    let charge = base_charge + payment_processing_cost;

    let markup_percentage = (100.0 + config.markup_percentage) / 100.0;

    let total_to_charge = (charge * markup_percentage).ceil();
    ChargeAmount {
        total_to_charge,
        with_vat: total_to_charge * 1.2,
    }
}

File: app/src/main/java/com/benjaminsproule/costcalculator/calculator/StripeCalculator.kt
package com.benjaminsproule.costcalculator.calculator

import com.benjaminsproule.costcalculator.store.Config
import kotlin.math.ceil

enum class StripeLocation(override val label: String) : Option {
    Local("Local"),
    EU("EU"),
    International("International"),
}

data class StripeSale(
    override val cost: Float,
    override val deliveryCosts: Float,
    val location: StripeLocation
) : Sale()

data class StripeCharge(
    override val numberOfMinutes: Float,
    override val materialCosts: List<Material>,
    override val deliveryCosts: Float,
    val location: StripeLocation
) : Charge()

class StripeCalculator(private val config: Config) : Calculator<StripeSale, StripeCharge> {
    companion object {
        const val PAYMENT_PROCESSING_PERCENTAGE: Float = 0.015f;
        const val EU_PAYMENT_PROCESSING_PERCENTAGE: Float = 0.025f;
        const val INTERNATIONAL_PAYMENT_PROCESSING_PERCENTAGE: Float = 0.035f;
        const val PAYMENT_PROCESSING_FEE: Float = 0.2f;
    }

    override fun basedOnSale(sale: StripeSale): SaleBreakdown {
        val saleTotal = sale.cost + sale.deliveryCosts;
        val paymentProcessingPercentage = if (sale.location == StripeLocation.EU) {
            EU_PAYMENT_PROCESSING_PERCENTAGE
        } else if (sale.location == StripeLocation.International) {
            INTERNATIONAL_PAYMENT_PROCESSING_PERCENTAGE
        } else {
            PAYMENT_PROCESSING_PERCENTAGE
        };
        val paymentProcessingCost = (saleTotal * paymentProcessingPercentage) + PAYMENT_PROCESSING_FEE;
        val tax = sale.cost * (config.taxRate / 100.0f);
        val revenue = sale.cost - paymentProcessingCost - tax;
        val percentageKept = (revenue / sale.cost) * 100.0f;
        val maxWorkingHours = revenue / config.hourlyRate;
        return SaleBreakdown(
            sale = sale.cost,
            deliveryCosts = sale.deliveryCosts,
            transactionCost = 0.0f,
            paymentProcessingCost = paymentProcessingCost,
            offsiteAdsCost = 0.0f,
            regulatoryOperatingFee = 0.0f,
            vatPaidByBuyer = 0.0f,
            vatOnSellerFees = 0.0f,
            totalFees = paymentProcessingCost,
            totalFeesWithVat = paymentProcessingCost,
            tax = tax,
            revenue = revenue,
            percentageKept = percentageKept,
            maxWorkingHours = maxWorkingHours,
        )
    }

    override fun howMuchToCharge(
        charge: StripeCharge
    ): ChargeAmount {
        val totalMaterialCosts = charge.materialCosts.map { material -> material.value }.sum();
        val baseCharge =
            ((charge.numberOfMinutes / 60.0f) * config.hourlyRate) + totalMaterialCosts + charge.deliveryCosts
        val paymentProcessingPercentage = if (charge.location == StripeLocation.EU) {
            EU_PAYMENT_PROCESSING_PERCENTAGE
        } else if (charge.location == StripeLocation.International) {
            INTERNATIONAL_PAYMENT_PROCESSING_PERCENTAGE
        } else {
            PAYMENT_PROCESSING_PERCENTAGE
        };
        val paymentProcessingCost = (baseCharge * paymentProcessingPercentage) + PAYMENT_PROCESSING_FEE;
        val chargeWithFees = baseCharge + paymentProcessingCost;

        val markupPercentage = (100.0f + config.markupPercentage) / 100.0f;

        val totalToCharge = ceil(chargeWithFees * markupPercentage);
        return ChargeAmount(
            totalToCharge = totalToCharge,
            withVat = totalToCharge * 1.2f,
        )
    }
}
